library/crypto.c

```diff
@@ -0,0 +1,179 @@
+#define DFL_SUBJECT_NAME        "C=US,O=NetFoundry,CN="
+#define DFL_KEY_USAGE           0
+#define DFL_FORCE_KEY_USAGE     0
+#define DFL_NS_CERT_TYPE        0
+#define DFL_FORCE_NS_CERT_TYPE  0
+#define DFL_MD_ALG              MBEDTLS_MD_SHA256
+
+struct csroptions
+{
+    const char *filename;       /* filename of the key file             */
+    const char *password;       /* password for the key file            */
+    int debug_level;            /* level of debugging                   */
+    const char *output_file;    /* where to store the constructed key file  */
+    const char *subject_name;   /* subject name for certificate request */
+    unsigned char key_usage;    /* key usage flags                      */
+    int force_key_usage;        /* Force adding the KeyUsage extension  */
+    unsigned char ns_cert_type; /* NS cert type                         */
+    int force_ns_cert_type;     /* Force adding NsCertType extension    */
+    mbedtls_md_type_t md_alg;   /* Hash algorithm used for signature.   */
+} csropt;
+
+
+int gen_csr(enroll_cfg *cfg) {
+
+    int ret = 1;
+    mbedtls_pk_context key;
+    mbedtls_ctr_drbg_context ctr_drbg;
+    char buf[1024];
+    mbedtls_entropy_context entropy;
+    const char *pers = "gen_csr";
+
+    // Set to sane values
+    mbedtls_x509write_csr_init( &cfg->x509_csr );
+    mbedtls_pk_init( &key );
+    mbedtls_ctr_drbg_init( &ctr_drbg );
+    memset( buf, 0, sizeof( buf ) );
+
+    csropt.filename            = DFL_FILENAME;
+    csropt.password            = DFL_PASSWORD;
+    csropt.debug_level         = DFL_DEBUG_LEVEL;
+    csropt.output_file         = DFL_OUTPUT_FILENAME;
+
+    char subject_name[256];
+    strcpy(subject_name, DFL_SUBJECT_NAME);
+    strcat(subject_name, cfg->zej->subject);
+    csropt.subject_name        = subject_name;
+
+    csropt.key_usage           = DFL_KEY_USAGE;
+    csropt.force_key_usage     = DFL_FORCE_KEY_USAGE;
+    csropt.ns_cert_type        = MBEDTLS_X509_NS_CERT_TYPE_SSL_CLIENT;
+    csropt.force_ns_cert_type  = 1;
+    csropt.md_alg              = MBEDTLS_MD_SHA256;
+
+    mbedtls_x509write_csr_set_md_alg( &cfg->x509_csr, csropt.md_alg );
+
+    mbedtls_x509write_csr_set_key_usage( &cfg->x509_csr, csropt.key_usage );
+
+    mbedtls_x509write_csr_set_ns_cert_type( &cfg->x509_csr, csropt.ns_cert_type );
+
+    mbedtls_entropy_init( &entropy );
+    if( ( ret = mbedtls_ctr_drbg_seed( &ctr_drbg, mbedtls_entropy_func, &entropy, (const unsigned char *) pers, strlen( pers ) ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_ctr_drbg_seed returned %d", ret);
+        return ZITI_CSR_GENERATION_FAILED;
+    }
+
+    if( ( ret = mbedtls_x509write_csr_set_subject_name( &cfg->x509_csr, csropt.subject_name ) ) != 0 ) {
+        ZITI_LOG(ERROR, "mbedtls_x509write_csr_set_subject_name returned %d", ret);
+        return ZITI_CSR_GENERATION_FAILED;
+    }
+
+    mbedtls_x509write_csr_set_key( &cfg->x509_csr, &cfg->pk_context );
+
+    memset( cfg->x509_CSR, 0, 4096 );
```

Comment:
you should probably use `sizeof(cfg->x509_CSR)` (or a macro) for the buffer size here.

Reply:
fixed.